GlobalBuildInfo plugin should search packed references for commit IDs

williamrandolph opened the pull request:
After a git garbage collection, I was unable to run `./gradlew` commands.

My failures were as follows:

```

FAILURE: Build failed with an exception.

* Where:
Build file '/Users/wbrafford/work/repos/elasticsearch/build.gradle' line: 37

* What went wrong:
An exception occurred applying plugin request [id: 'elasticsearch.global-build-info']
> Failed to apply plugin [id 'elasticsearch.global-build-info']
   > unable to read the git revision

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 1s
```

If you have version 2.23.0 of git, you should be able to reproduce after running:

```
git pack-refs --all
```

Documentation on `git-pack-refs` is here: https://git-scm.com/docs/git-pack-refs

This code needs a couple of improvements (for example, there's no check for the existence of the packed-refs file), but I'm going ahead and posting the PR for discussion.

williamrandolph committed:
GlobalBuildInfo plugin searches packed references

In recent versions of Git, references may be packed in a packed-refs
file. If this happens, Gradle will need to look in that file to find
build information.

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Build)

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/info/GlobalBuildInfoPlugin.java:
```diff
@@ -248,7 +250,20 @@ public static String gitRevision(File rootDir) {
             final String ref = readFirstLine(head);
             if (ref.startsWith("ref:")) {
-                revision = readFirstLine(gitDir.resolve(ref.substring("ref:".length()).trim()));
+                String branchname = ref.substring("ref:".length()).trim();
```
Sorry for being pedantic but this threw me off when I was digging through the git file structure. This is actually the name of the reference (ex: `refs/heads/master`) not simply the name of the branch. Can we rename this `refName` or similar to account for this?

williamrandolph replied:
That's the good kind of pedantic; you're absolutely right. I'll do it.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/info/GlobalBuildInfoPlugin.java:
```diff
@@ -248,7 +250,20 @@ public static String gitRevision(File rootDir) {
+                    revision = readFirstLine(refFile);
+                } else {
+                    // Check packed references for commit ID
+                    Pattern p = Pattern.compile("^([a-f0-9]+) " + branchname + "$");
```
We could make this even more strict since we know a sha1 hash is 40 characters with `[a-f0-9]{40}`.

mark-vieira replied:
Also, do we know what the entries that look like `^6d5d330548f66c1ba6d2913737ba93c5d5a39de9` mean? I couldn't find any documentation describing the actual format of this file.

williamrandolph replied:
Looking at the last lines of my `.git/packed-refs` file…
```
d7d5b3fbc408e3b4060d9df33dec9c43e29e9cbb refs/tags/v7.4.0
^22e1767283e61a198cb4db791ea66e3f11ab9910
```
It looks like we get that carrot notation when Elasticmachine tags a release. I'm not sure what the specific terminology is, but `22e17672` is the commit that's tagged for 7.4 and `git show d7d5b3fbc` shows the tag. Perhaps the first line represents the "annotated tag object"? I wouldn't want to try to get any more definitive without finding the format documentation.

mark-vieira replied:
Just curious mostly. Seems we can safely ignore those entries.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/info/GlobalBuildInfoPlugin.java:
```diff
@@ -248,7 +250,20 @@ public static String gitRevision(File rootDir) {
+                } else {
+                    // Check packed references for commit ID
+                    Pattern p = Pattern.compile("^([a-f0-9]+) " + branchname + "$");
+                    revision = Files.lines(dotGit.resolve("packed-refs"))
```
We should create this stream in a try-with-resources block to ensure we properly close it. In fact, we should refactor `readFirstLine()` below to do the same here as the Gradle daemon could be unintentionally holding a bunch of file handles here.

```
try  (Stream<String> lines = Files.lines(dotGit.resolve("packed-refs")) {
    // do stuff
}
```

Doc of `Files.lines()`:
```
     * This method must be used within a try-with-resources statement or similar
     * control structure to ensure that the stream's open file is closed promptly
     * after the stream's operations have completed.
```

mark-vieira reviewed (approved):
Couple of comments, otherwise LGTM. Thanks for catching this @williamrandolph.

williamrandolph committed:
Close file references and clean up variable names

The Files.readLines method needs to be used in a try-with-resources
block. Also, we can be more specific in the regex, and what was the
"branchname" variable should really be called "refName".

williamrandolph commented:
@elasticmachine Please run elasticsearch-ci/2

mark-vieira commented:
@williamrandolph I think this is in a state where it can be merged once the conflicts are resolved. Let's not let this branch get any more stale than it is.

williamrandolph committed:
Merge changes from master

williamrandolph committed:
Restore import dropped during merge

williamrandolph commented:
@elasticmachine please run elasticsearch-ci/packaging-sample-matrix

williamrandolph commented:
@elasticmachine please run elasticsearch-ci/2
